File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) 
    {
        if(head == NULL || head->next == NULL)
        {
            return NULL;
        }    
        ListNode *slow = head;
        ListNode *fast = head;

        while(fast!=NULL && fast->next!=NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
            
            if(fast == slow)
            {
                return detectNode(head, fast);
            }
            
        }
        return NULL;
    }
    ListNode *detectNode(ListNode *head, ListNode *fast)
    {
        ListNode *curr = head;
        ListNode *temp = fast;

        if(curr == temp)
        {
            return fast;
        }
        else
        {
            while(curr!=temp)
            {
                curr = curr->next;
                temp = temp->next;
            }
        }
        return curr;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        
        ListNode fast = head;
        ListNode slow = head;
        
        // Phase 1: Detect if cycle exists
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            
            if (fast == slow) {
                // Cycle detected! Now find the start
                // Phase 2: Find cycle start
                ListNode entry = head;
                while (entry != slow) {
                    entry = entry.next;
                    slow = slow.next;
                }
                return entry;  // or return slow, they're at same position
            }
        }
        
        return null;  // No cycle
    }
}